lib/grp.gi

```diff
@@ -783,16 +805,35 @@ InstallMethod( DerivedSeriesOfGroup,
     D := DerivedSubgroup( G );
    
     while
-      # we don't know that the group has no generators
-      (not HasGeneratorsOfGroup(S[Length(S)]) or
-	    Length(GeneratorsOfGroup(S[Length(S)]))>0) and
-      ( (not HasAbelianInvariants(S[Length(S)]) and D <> S[ Length(S) ]) or
-	    Length(AbelianInvariants(S[Length(S)]))>0) do
+      (not HasIsTrivial(S[Length(S)]) or
+	    not IsTrivial(S[Length(S)])) and
+      (
+        (not HasIsPerfectGroup(S[Length(S)]) and
+         not HasAbelianInvariants(S[Length(S)]) and D <> S[ Length(S) ]) or
+        (HasIsPerfectGroup(S[Length(S)]) and not IsPerfectGroup(S[Length(S)]))
+        or (HasAbelianInvariants(S[Length(S)])
+                            and Length(AbelianInvariants(S[Length(S)])) > 0)
+      ) do
         Add( S, D );
         Info( InfoGroup, 2, "DerivedSeriesOfGroup: step ", Length(S) );
         D := DerivedSubgroup( D );
     od;
 
+    # set filters if the last term is known to be trivial
+    if HasIsTrivial(S[Length(S)]) and IsTrivial(S[Length(S)]) then
+      SetIsSolvableGroup(G, true);
+      if Length(S) <=2 then
+        Assert(2, IsAbelian(G));
+        SetIsAbelian(G, true);
+      fi;
+    fi;
+
+    # set IsAbelian filter if length of derived series is more than 2
+    if Length(S) > 2 then
+      Assert(2, not IsAbelian(G));
+      SetIsAbelian(G, false);
+    fi;
+
```

Comment:
How about simplifying this chunk similar to what I proposed for `IsSolvableGroup` ?